src/dune_rules/install_rules.ml

```diff
@@ -412,10 +420,18 @@ end = struct
           ~dir
           ~relative_dst_path_starts_with_parent_error_when:`Deprecation_warning_from_3_11
       in
-      Memo.List.map dirs_expanded ~f:(fun fb ->
+      Memo.List.concat_map dirs_expanded ~f:(fun fb ->
         let loc = File_binding.Expanded.src_loc fb in
-        let+ entry = make_entry ~kind:`Directory fb in
-        Install.Entry.Sourced.create ~loc entry)
+        let* entry = make_entry ~kind:`Directory fb in
+        let* rule_opt = Load_rules.get_rule (Path.build entry.src) in
+        let+ entries =
+          match rule_opt with
+          | None -> Memo.return [ entry ]
+          | Some rule ->
+            let+ { targets; _ } = Build_system.execute_rule rule in
```

Comment:
There's a dependency loop here: the set of installed files is necessary to populate the contents of the install file; but dune-package is also installed so we can't easily get the set of installed files when creating the contents of `dune-package`. I think we could try to make `gen_install_file` in this module generate `dune-project` as well but I'm not 100% sure. Anyway, if I understand correctly we're going to execute this rule just a bit later so I don't think it's too bad to call `execute_rule` directly here to get the produced targets.

Reply:
There's another problem as well. With the current code, we now need to build all directory targets to just generate the rule for the dune-package file. This has two practical disadvantages:

1. Depending on `(package foo)` is going to build the directory targets of all other packages. 
2. It will make the implementation of https://github.com/ocaml/dune/issues/8652 more difficult for similar reasons. We'll now won't be able to list the entries of a package without building things.

So I'm afraid this doesn't work. I think we should bite the bullet and just add `(dirs )` to dune-package.

Reply:
ok. I did so in #8953